core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala

```diff
@@ -160,11 +160,15 @@ class HadoopMapReduceCommitProtocol(
 
     val taskAttemptContext = new TaskAttemptContextImpl(jobContext.getConfiguration, taskAttemptId)
     committer = setupCommitter(taskAttemptContext)
-    committer.setupJob(jobContext)
+    if (!dynamicPartitionOverwrite) {
```

Comment:
We need to add comments to explain it. It looks to me that the hadoop output committer doesn't support concurrent writing to the same directory by design, so there is nothing we can do at Spark side.

The fix here is to avoid using the hadoop output committer when `dynamicPartitionOverwrite=true`. I'm fine with this fix.

BTW, when writing partitioned table with `dynamicPartitionOverwrite=false`, can we support it as well?

Reply:
also cc @advancedxy

Reply:
and for non-partitioned table, can we clean up the staging dir when the job is killed?

Reply:
@advancedxy  has discussed with me offline about writing partitioned table with dynamicPartitionOverwrite=false .
He propose a suggestion that, we can add output existence check when dynamicPartitionOverwrite=false.

Reply:
For a non-partitioned table, dynamicPartitionOverwrite is false,  and the staging dir is under JobAttemptPath(_temporary/0), I think the staging dir will be cleaned up by FileOutputCommitter.abortJob().

Reply:
> I think the staging dir will be cleaned up by FileOutputCommitter.abortJob().

Why it can't be cleaned when `dynamicPartitionOverwrite=true`?

Reply:
When a job is killed, its staging dir can be cleaned up by `abortJob` method.
But when an application is killed, its job's staging dir would not be cleaned up gracefully.

Reply:
> > I think the staging dir will be cleaned up by FileOutputCommitter.abortJob().
> 
> Why it can't be cleaned when `dynamicPartitionOverwrite=true`?

For the case in PR description,  It is happened when appA(static partition overwrite) is killed and its staging dir is not cleaned up gracefully, then appB commits parts result of appA.

Reply:
OK, so we can't rely on the job cleanup. And ideally we should use different staging dir for each job.

That said, seems we can't fix the problem for non-partitioned table if we continue to use the hadoop output committer.

Reply:
Yes,  a solution proposed by advancedxy is to add job attempt path existence check  for non-partitioned table 

And the implement of `InsertIntoHiveTable` uses different staging dir for each job.